src/d3d9/d3d9_bridge.h

```diff
@@ -0,0 +1,121 @@
+#pragma once
+
+#include <windows.h>
+#include "../util/config/config.h"
+#include "../util/util_error.h"
+#include "../util/log/log.h"
+
+#include "../vulkan/vulkan_loader.h"
+
+#include <bitset>
+
+// NOTE: You must include "d3d9_include.h" or "d3d8_include.h" before this header.
+
+/**
+ * The D3D9 bridge allows D3D8 to access DXVK internals.
+ * For Vulkan interop without needing DXVK internals, see d3d9_interop.h.
+ */
+
+namespace dxvk {
+
+  class D3D9DeviceEx;
+  class D3D9InterfaceEx;
+
+#if defined(_MSC_VER)
+  class DECLSPEC_UUID("D3D9ACAB-42A9-4C1E-AA97-DEADFADED420") D3D9Bridge;
+  class DECLSPEC_UUID("D3D9ACAB-A407-773E-18E9-CAFEBEEF2000") D3D9InterfaceBridge;
+#endif
+
+  class D3D9Bridge : public IUnknown {
+
+  // D3D8 keeps D3D9 objects contained in a namespace.
+  #ifdef DXVK_D3D9_NAMESPACE
+    using IDirect3DSurface9 = d3d9::IDirect3DSurface9;
+  #endif
+
+  public:
+    using Type = D3D9DeviceEx;
+
+    D3D9Bridge(D3D9DeviceEx* pDevice) : m_device(pDevice) {}
+    virtual ~D3D9Bridge() = default;
+
+    virtual ULONG STDMETHODCALLTYPE AddRef();
+    virtual ULONG STDMETHODCALLTYPE Release();
+    virtual HRESULT STDMETHODCALLTYPE QueryInterface(
+            REFIID  riid,
+            void** ppvObject);
+
+    virtual void SetAPIName(const char* name);
+
+    virtual void AddBatchCalls(uint32_t count);
+
+    virtual void SetShadowBuffersEnabled(bool enabled);
+
+    virtual void ForceEndScene();
+
+    virtual HRESULT UpdateTextureFromBuffer(
+        IDirect3DSurface9*        pDestSurface,
+        IDirect3DSurface9*        pSrcSurface,
+        const RECT*               pSrcRect,
+        const POINT*              pDestPoint);
+
+
+    // Indicate that D3D9 shouldn't complain about an unhandled
+    // D3DRENDERSTATE because it's implemented elsewhere.
+    template <typename... Args>
+    inline void AddSupportedRenderStates(Args... state) {
+      ((m_supportedRenderStates[state] = true), ...);
+    }
+
+    inline void RenderStateNotSupported(uint8_t renderState) {
+      m_supportedRenderStates[renderState] = false;
+    }
+
+    constexpr bool IsRenderStateSupported(uint8_t renderState) const {
+      return m_supportedRenderStates[renderState];
+    }

```

Comment:
I don't see the point of this if it's just to avoid log spam, those are only logged once. Shouldn't you be catching them on the D3D8 side anyway? Why are you forwarding unsupported stuff to us?